### PR TITLE
IGNITE-14924 HOCON configuration source and representation implemented.

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-14924

### CODE REVIEW

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.List;
+import com.typesafe.config.ConfigObject;
+import com.typesafe.config.ConfigValue;
+import com.typesafe.config.impl.ConfigImpl;
+import org.apache.ignite.internal.configuration.ConfigurationRegistry;
+import org.apache.ignite.internal.configuration.tree.ConfigurationSource;
+
+public class HoconConverter {
+    /**
+     * Converts configuration subtree to HOCON {@link ConfigValue} instance.
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java:29]:
```suggestion
     * Converts configuration subtree to a HOCON {@link ConfigValue} instance.
```

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @param registry Configuration registry instance.
+     * @param path Path to configuration subtree. Can be empty, can't be {@code null}.
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java:32]:
```suggestion
     * @param path Path to the configuration subtree. Can be empty, can't be {@code null}.
```

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+     * @return {@link ConfigValue} instance that represents configuration subtree.
+     * @throws IllegalArgumentException If {@code path} is not found in current configuration.
+     */
+    public static ConfigValue represent(ConfigurationRegistry registry, List<String> path) throws IllegalArgumentException {
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java:36]:
```suggestion
    public static ConfigValue represent(ConfigurationRegistry registry, @NotNull List<String> path) throws IllegalArgumentException {
```

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/NamedListNode.java [hunk 4]
```diff
@@ -43,9 +47,12 @@
     /**
      * Default constructor.
      *
+     * @param syntheticKey Name of the synthetic configuration value that will represent keys in special ordered
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/NamedListNode.java:50]:
```suggestion
     * @param syntheticKey Name of the synthetic configuration value that will represent keys in a specially ordered
```

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/NamedListNode.java [hunk 5]
```diff
@@ -43,9 +47,12 @@
...
+     *                     representation syntaxis.
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/NamedListNode.java:51]:
```suggestion
     *                     representation syntax.
```

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconPrimitiveConfigurationSource.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                return clazz.cast(numberValue.floatValue());
+            else if (clazz == Double.class)
+                return clazz.cast(numberValue.doubleValue());
+        }
+
+        throw new IllegalArgumentException("Unsupported type: " + clazz);
+    }
+
+    /**
+     * Checks that a number fits into the given bounds.
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconPrimitiveConfigurationSource.java:152]: Params and return are missing

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/NamedListNode.java (hunk 5 above)
Comment (SammyVimes) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/NamedListNode.java:51]: Not sure if it's clear enough. Maybe rewrite it somehow?

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    /**
+     * @param hoconCfg HOCON that has to be converted to the configuration source.
+     * @return HOCON-based configuration source.
+     */
+    public static ConfigurationSource hoconSource(ConfigObject hoconCfg) {
+        return new HoconObjectConfigurationSource(null, List.of(), hoconCfg);
+    }
+
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconConverter.java:49]: Done

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconPrimitiveConfigurationSource.java (hunk 6 above)
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconPrimitiveConfigurationSource.java:152]: Yeah, they are missing on many methods here

Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconPrimitiveConfigurationSource.java:161]: Done

**Review by SammyVimes — APPROVED** [review]
LGTM!

**Pull request merged**